mmcv/runner/ipu_runner.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from abc import ABCMeta
+import torch
+
+from .iter_based_runner import IterBasedRunner
+from .epoch_based_runner import EpochBasedRunner
+from .builder import RUNNERS
+from .hooks import HOOKS
+
+from mmcv.runner.ipu import IPU_MODE
+if IPU_MODE:
+    from mmcv.runner.ipu import (parse_ipu_options,
+                                 build_from_cfg_with_wrapper, IPU_MODE,
+                                 ipu_model_wrapper, wrap_optimizer_hook,
+                                 IPUFp16OptimizerHook, wrap_lr_update_hook,)
+
+
+class IPUBaseRunner(metaclass=ABCMeta):
+    """A base runner for IPU device, should be inherited with
+    base_runner.BaseRunner
+    This runner has some extra processes for IPU which are showed below:
+    1. Parse options for IPU
+    2. wrap pytorch model for IPU
+    3. Raise errors while encountering illegal ussage
+    4. Input IPU options and initialize dataloader if finding a instance of
+       of IPUDataloader
+
+    Args:
+        model (pytorch.model)
+        options (mmcv.Config, dict): Options that will be used to compile
+            and run the model.
+        modules_to_record (mmcv.Config, list): idx or name of modules which
+            will be recorded for output. It is necessary to specify output for
+            static graph of model training or inference.
+        ipu_model_cfg (mmcv.Config, dict): config of model partition and
+            recomputing checkpoint
+        fp16_cfg (mmcv.Config): config for fp16 training
+        batch_processor (callable): A callable method that process a data
+            batch. Should be None for IPU runner
+        kwargs: other kwargs please check Class base_runner.BaseRunner
+    """
+    def __init__(
+            self,
+            model,
+            ipu_options={},
+            modules_to_record=[],
+            ipu_model_cfg={},
```

Comment:
Using empty list or dict as default values is unsafe.

Reply:
Is it because of recursion or multithreading？